crates/provider-wit-bindgen/src/lib.rs

```diff
@@ -41,187 +55,926 @@ const EXPORTS_MODULE_NAME: &str = "exports";
+#[derive(Debug, Default, PartialEq, Eq, Clone)]
+enum WitFunctionLatticeTranslationStrategy {
+    /// Automatically determine how to bundle a function and it's arguments to be sent over the
+    ///
+    /// Generally, this means detecting whether the function has a single argument or multiple
+    /// and choosing whether to send that single argument or bundle up all arguments into a generated Invocation struct.
+    #[default]
+    Auto,
+
+    /// First argument usage assumes that every function that should be exported to the lattice
+    /// has *one* argument, which is the object that should be serialized and set out onto the lattice.
+    ///
+    /// For example, the following WIT:
+    ///
+    /// ```ignore
+    /// interface example {
+    ///   f: func(input: string);
+    ///
+    ///   record g-request {
+    ///     first: string;
+    ///     second: u32;
+    ///     third: bool;
+    ///   }
+    ///
+    ///   g: func(input: g-request);
+    ///
+    ///   h: func(first: bool, second: string);
+    /// }
+    /// ```
+    ///
+    /// Under this setting, bindgen will produce an error on function `h`, as it contains multiple arguments
+    FirstArgument,
+
+    /// Argument bundling assumes that every function that should be exported to the lattice
+    /// has *one or more* arguments, which are bundled into an object that should be serialized and set out onto the lattice.
+    ///
+    /// For example, the following WIT:
+    ///
+    /// ```ignore
+    /// package examples:arg-bundle;
+    ///
+    /// interface example {
+    ///   f: func(input: string);
+    ///
+    ///   record g-request {
+    ///     first: string;
+    ///     second: u32;
+    ///     third: bool;
+    ///   }
+    ///
+    ///   g: func(input: g-request);
+    ///
+    ///   h: func(first: bool, second: string);
+    /// }
+    /// ```
+    ///
+    /// Under this setting, bindgen will not produce an error on any function, but will be slightly
+    /// inefficient as it will wrap `f` in a generated `ExamplesArgBundleFInvocation` struct (containing one member, `input`).
+    BundleArguments,
+}
+
+impl WitFunctionLatticeTranslationStrategy {
+    /// Translate a wit-bindgen generated trait function for use across the lattice
+    fn translate_import_fn_for_lattice(
+        &self,
+        wit_iface_name: WitInterfaceName,
+        trait_method: &TraitItemMethod,
+        struct_lookup: &StructLookup,
+        type_lookup: &TypeLookup,
+        bindgen_cfg: &ProviderBindgenConfig,
+    ) -> anyhow::Result<(WitInterfaceName, LatticeMethod)> {
+        // Calculate the lattice method name
+        let lattice_method_name = LitStr::new(
+            match (
+                bindgen_cfg.contract.as_str(),
+                &bindgen_cfg.wit_ns,
+                &bindgen_cfg.wit_pkg,
+            ) {
+                // To facilitate backwards compatibility with existing wasmcloud:keyvalue 
+                ("wasmcloud:keyvalue", _, _) => {
```

Comment:
Is there a way we can avoid hardcoding interface names here?
At the very least we should have a detailed explanation here as to why we *have to* hardcode and cannot generate this

Reply:
It's not robust to force the user to hard-code a certain package, interface, or other name inside their WIT, and generating this from that.

The thing we have control over is the wasmcloud contract -- and that's the thing that actually forces the need for backwards compatibility. 

Also note that the that is the *wasmcloud contract* that we're checking.

Reply:
> It's not robust to force the user to hard-code a certain package

Could you expand on this point to help me understand the reasoning here?

Generating `KeyValue` namespace from `key-value` WIT interface name seems to be very robust and clear to me.
What names do other cases generate? (do note that we need to generate namespace from *interface* name, not package)

Reply:
WIT interface naming can be arbitrary -- if some other contract named an interface `key-value` (let's say someone was writing a `theircompany:nats` provider, and `key-value` was the name of NATS KV functionality), they'd run into this special case as well, right? This is *especially* the case if we don't take into account the contract name -- any interface named `key-value`.

We just happen to have `key-value` all in one interface stuffed in `keyvalue.wit` *right now* but there's no guarantee that is the case for any future WITs, and arguably it's not good practice to stuff unrelated things in the same interface to begin with.

Arguably, `set-*` functions should be in a `set` interface, `list` operations in `list`, ETC, similar to the best practices used by `wasi:keyvalue` and other contracts that try to group functionality. Keying off of interface name unnecessarily restricts the structure of the WIT, forcing us to make one big all-in-one interface.

What determines the lattice method prefix here is  the wasmcloud contract (in this case `wasmcloud:keyvalue`), not the WIT interface name. 

One thing that could work is using these two in combination, but that still causes unnecessary requirements on the WIT writers.

Is there a use-case you're envisioning in particular around using the interface name, or is it simply to avoid the match statement (which is necessary anyway, because we must check for `"key-value"`, if I'm understanding your suggestion correctly)

Reply:
The only way to uniquely identify methods is to use a combination of package name, interface name and method. There's no other way to disambiguate - for a real example, see https://github.com/WebAssembly/wasi-clocks/blob/30af30126ffe85455d06eb29a11d99f2519b6c3d/wit/monotonic-clock.wit#L21 and https://github.com/WebAssembly/wasi-clocks/blob/30af30126ffe85455d06eb29a11d99f2519b6c3d/wit/wall-clock.wit#L35

Contrary to your example, including the interface name would, in fact, allow us to disambiguate these two, but not the package name.

At present, links are defined per-package, so for the overwhelming majority of cases, identifying methods using interface name/method name pair should be good enough, since providers would rarely implement interfaces from more than one package, in fact, if we passed the package name from the linkdef into the call, we could have all 3 pieces required to uniquely identify the method being called (and it may actually be already how it works, I'd need to double-check, perhaps @thomastaylor312 could comment here)

I am not envisioning any match statements here, but rather interface name to be converted to PascalCase as outlined here: https://github.com/wasmCloud/wasmCloud/pull/672#issue-1909373051.
So `foo` would turn into `Foo`, and `foo-bar` would turn into `FooBar`. This would match existing naming scheme and align with existing ecosystem letting us smoothly transition.

Reply:
I was having a bit of a hard time following along, but I think I agree with @rvolosatovs here. FWIW, I consider all of the providers we'll be updating to be a breaking change, so we don't have to perfectly match the old `wasmcloud:keyvalue` syntax. The wasifill generator stuff I'm doing is things like `my:package/interface/PascalCaseFunctionName` (I was going to come back and match with whatever you do here) when sending an invocation across the wire.

Reply:
Thanks for weighing in Taylor -- this isn't about the general case (there's [code that does something similar](https://github.com/wasmCloud/wasmCloud/pull/662/files#diff-c35e6db0016f48f0c6fe14869e51b1fbffd9c6d8db1be7933222b5d519e6045aR337) to what you did). 

This is about how we generate the across-the-lattice method/operation name (ex. `KeyValue.Contains`). 

The WIT that we have right now looks like this:

```
package wasmcloud:keyvalue

interface key-value {
    // ... stuff 
    contains: func(input: string) -> bool
}
```

From what I can understand, Roman wants to use the `interface key-value` part + `contains` to generate `KeyValue.Contains`. This means that anyone who uses this bindgen has to name their interface `key-value` (/`messaging`/`sqldb-postgres`/...) and that when bindgen sees `key-value`, it should convert that. 

This means that no one can use the name `key-value` or `messaging` without triggering this functionality. 

What I'm proposing here is taking the contract name (which we need to know/is an input for a reason), and using the method name on the exposed interface to figure it out. 

So if you know you're targeting the `wasmcloud:keyvalue` contract and the on-lattice prefix is `KeyValue.`, you use that -- along with the exposed function name, upcased -- `KeyValue.Contains`.

It actually *has* to be `KeyValue.Contains` for backwards compatibility with the existing providers (unless something in the plan has changed), this is the specific case of already existing stuff

Reply:
After chatting, we're going to just convert interface to PascalCase here and add on the method name

- [ ] convert interface to PascalCase and use for operation name